benchmark/Streamly/Benchmark/Data/Serialize.hs

```diff
@@ -0,0 +1,455 @@
+    {-# INLINE (==) #-}
+    (==)
+        (Product25 a1 a2 a3 a4 a5 a6 a7 a8 a9 a10 a11 a12 a13 a14 a15 a16 a17 a18 a19 a20 a21 a22 a23 a24 a25)
+        (Product25 b1 b2 b3 b4 b5 b6 b7 b8 b9 b10 b11 b12 b13 b14 b15 b16 b17 b18 b19 b20 b21 b22 b23 b24 b25)
+           =   a1  == b1
+            && a2  == b2
+            && a3  == b3
+            && a4  == b4
+            && a5  == b5
+            && a6  == b6
+            && a7  == b7
+            && a8  == b8
+            && a9  == b9
+            && a10 == b10
+            && a11 == b11
+            && a12 == b12
+            && a13 == b13
+            && a14 == b14
+            && a15 == b15
+            && a16 == b16
+            && a17 == b17
+            && a18 == b18
+            && a19 == b19
+            && a20 == b20
+            && a21 == b21
+            && a22 == b22
+            && a23 == b23
+            && a24 == b24
+            && a25 == b25
+
+-------------------------------------------------------------------------------
+
+-- Simple non-recursive ADT
+data CustomDT1
+    = CDT1C1
+    | CDT1C2 Int
+    | CDT1C3 Int Int
+    deriving (Generic, Show, Eq)
+
+#ifndef USE_TH
+instance SERIALIZE_CLASS CustomDT1
+#else
+$(DERIVE_CLASS ''CustomDT1)
+#endif
+
+-------------------------------------------------------------------------------
+
+{-# INLINE getSize #-}
+getSize :: forall a. SERIALIZE_CLASS a => a -> Int
+#ifdef USE_UNBOX
+getSize _ = sizeOf (Proxy :: Proxy a)
+#else
+getSize val =
+    case size :: Size a of
+        ConstSize x -> x
+        VarSize f -> f val
+#endif
+
+#ifndef USE_UNBOX
+{-# INLINE sizeOfOnce #-}
+sizeOfOnce :: forall a. Serialize a => a -> Int
+sizeOfOnce val = getSize val
+
+-- Recursive ADT
+data BinTree a
+  = Tree (BinTree a) (BinTree a)
+  | Leaf a
+  deriving (Show, Read, Eq, Generic)
+
+-- XXX If a is ConstSize we can use an array to represent the binary tree. If
+-- the tree is balanced space wastage would be minimum.
+#ifndef USE_TH
+instance Serialize (BinTree a)
+#else
+$(deriveSerialize ''BinTree)
+#endif
+
+instance NFData a => NFData (BinTree a) where
+  rnf (Leaf a) = rnf a `seq` ()
+  rnf (Tree l r) = rnf l `seq` rnf r `seq` ()
+
+instance Arbitrary a => Arbitrary (BinTree a) where
+  arbitrary = oneof [Leaf <$> arbitrary, Tree <$> arbitrary <*> arbitrary]
+
+mkBinTree :: (Arbitrary a) => Int -> IO (BinTree a)
+mkBinTree = go (generate $ arbitrary)
+
+    where
+
+    go r 0 = Leaf <$> r
+    go r n = Tree <$> go r (n - 1) <*> go r (n - 1)
+
+{-
+prop :: Property
+prop = forAll arbitrary (ioProperty . test)
+
+    where
+
+    test :: BinTree Int -> IO Bool
+    test t = do
+        let n =
+                case size :: Size (BinTree Int) of
+                    ConstSize x -> x
+                    VarSize f -> f t
+        arr <- newBytes n
+        _ <- serialize 0 arr t
+        (_, t1) <- deserialize 0 arr
+        return $ t1 == t
+
+runQC :: IO ()
+runQC = quickCheckWith stdArgs {maxSuccess = 100} prop
+-}
+#endif
```

Comment:
Can remove this.

Reply:
This change is already made in another PR.